### PR TITLE
Update the functionality and performance of most methods.

### PR DESCRIPTION
**Pull request by Zakaria95Ahmed**
Update the functionality and performance of most methods.

### CODE REVIEW

File: product-images/ffa7601b-7584-47ee-9aff-9be0cee04940.png
Comment (tmohamed) [inline product-images/ffa7601b-7584-47ee-9aff-9be0cee04940.png:1]: Why product-images directory is included in git ?

File: src/main/java/com/fruits/ecommerce/configuration/SecurityConfig/JWT_Filters/JwtAuthorizationFilter.java [hunk 1]
```diff
@@ -15,25 +20,25 @@
...
 
-
 @Slf4j
 @Component
 public class JwtAuthorizationFilter extends OncePerRequestFilter {
+
     private final JWTTokenProvider jwtTokenProvider;
     private final SecurityProperties securityProperties;
-//    private final CustomUserDetailsService customUserDetailsService;
+    private final IUserService userService;
```
Comment (tmohamed) [inline src/main/java/com/fruits/ecommerce/configuration/SecurityConfig/JWT_Filters/JwtAuthorizationFilter.java:34]: Let's rename to UserService instead

File: src/main/java/com/fruits/ecommerce/configuration/SecurityConfig/SecurityCore/UserAuthorizationUtils.java [hunk 2]
```diff
@@ -0,0 +1,29 @@
+package com.fruits.ecommerce.configuration.SecurityConfig.SecurityCore;
+
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.security.core.Authentication;
+import org.springframework.stereotype.Component;
+
+@Slf4j
+@Component("userAuthorizationUtils")
```
Comment (tmohamed) [inline src/main/java/com/fruits/ecommerce/configuration/SecurityConfig/SecurityCore/UserAuthorizationUtils.java:8]: No need to define bean name like this, its automatically defined by Spring

File: src/main/java/com/fruits/ecommerce/controller/AuthController.java [hunk 3]
```diff
@@ -40,63 +42,175 @@ public ResponseEntity<UserDTO> registerUser(@Valid @RequestBody UserDTO userDTO)
...
+     * @param userDTO Updated user details.
+     * @return ResponseEntity with updated UserDTO.
+     */
+    // We use @PreAuthorize("#userId == authentication.principal.id and hasRole('ADMIN')")
+    // To Prevents users from changing each other's passwords
+    // Ensures that each user can only change their own password
+    // Security level is verified before accessing the code
+    // Can be easily changed without modifying the method code
+    @PutMapping("/{userId}")
+    @PreAuthorize("hasRole('ADMIN') or @userAuthorizationUtils.hasUserId(authentication, #userId)")
```
Comment (tmohamed) [inline src/main/java/com/fruits/ecommerce/controller/AuthController.java:150]: Why 'ADMIN' role alone is not enough for this method ?

File: src/main/java/com/fruits/ecommerce/controller/CartController.java [hunk 4]
```diff
@@ -26,9 +27,10 @@ public ResponseEntity<Void> addProductToCart(
         return ResponseEntity.status(HttpStatus.CREATED).build();
     }
     @GetMapping
-    @PreAuthorize("hasAnyRole('CUSTOMER','USER')")
+    @PostAuthorize("hasRole('CUSTOMER')")
```
Comment (tmohamed) [inline src/main/java/com/fruits/ecommerce/controller/CartController.java:30]: This role can be added to Security configuration instead for the path "**/cart/**"

File: src/main/java/com/fruits/ecommerce/models/dtos/CartItemDTO.java [hunk 5]
```diff
@@ -8,9 +8,7 @@
 @Setter
 @Getter
 public class CartItemDTO {
-    //    private ProductDTO productId;
```
Comment (tmohamed) [inline src/main/java/com/fruits/ecommerce/models/dtos/CartItemDTO.java:11]: Let's keep the old format instead, variable definition per line

File: src/main/java/com/fruits/ecommerce/models/dtos/PasswordChangeRequestDTO.java [hunk 6]
```diff
@@ -0,0 +1,19 @@
+package com.fruits.ecommerce.models.dtos;
+
+import jakarta.validation.constraints.NotBlank;
+import jakarta.validation.constraints.Size;
+import lombok.Getter;
+import lombok.Setter;
+
+@Getter
+@Setter
+public class PasswordChangeRequestDTO {
```
Comment (tmohamed) [inline src/main/java/com/fruits/ecommerce/models/dtos/PasswordChangeRequestDTO.java:10]: Let's rename to ChangePasswordRequestDTO instead

File: src/main/java/com/fruits/ecommerce/services/Interfaces/IUserService.java [hunk 7]
```diff
@@ -43,4 +47,42 @@ public interface IUserService {
     List<UserDTO> getAllCustomers() throws RoleNotFoundException;
     List<UserDTO> getAllUsers();
```
Comment (tmohamed) [inline src/main/java/com/fruits/ecommerce/services/Interfaces/IUserService.java:48]: Let's rename to UserService instead

File: src/main/resources/db/migration/V1__Create_users_and_roles_tables.sql [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+    address       VARCHAR(35),
+    created_at    TIMESTAMP    NOT NULL DEFAULT CURRENT_TIMESTAMP,
+    updated_at    TIMESTAMP    NOT NULL DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP,
+    last_login    TIMESTAMP,
+    is_active     BOOLEAN      NOT NULL DEFAULT TRUE,
+    is_not_locked BOOLEAN      NOT NULL DEFAULT TRUE
+);
+
+-- Add indexes
+CREATE INDEX idx_users_email ON users (email);
```
Comment (tmohamed) [inline src/main/resources/db/migration/V1__Create_users_and_roles_tables.sql:32]: Do we need index on 'email' ?

File: src/main/resources/db/migration/V3__Create_products_and_product_images_tables.sql [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+    PRIMARY KEY (id),
+    CONSTRAINT fk_product_images_product
+        FOREIGN KEY (product_id) REFERENCES products (id)
+            ON DELETE CASCADE
+) ENGINE = InnoDB
+  DEFAULT CHARSET = utf8mb4
+  COLLATE = utf8mb4_unicode_ci;
+
+-- Add indexes for performance
+CREATE INDEX idx_product_name ON products (name);
```
Comment (tmohamed) [inline src/main/resources/db/migration/V3__Create_products_and_product_images_tables.sql:35]: Do we need index on 'name' ?

**Review by tmohamed — CHANGES_REQUESTED** [review]
Thanks Eng.Zakaria for your great effort, would you please check comments on the PR

File: src/main/resources/db/migration/V1__Create_users_and_roles_tables.sql (hunk 8 above)
Comment (Zakaria95Ahmed) [reply src/main/resources/db/migration/V1__Create_users_and_roles_tables.sql:32]: email used in login operation

File: product-images/ffa7601b-7584-47ee-9aff-9be0cee04940.png
Comment (Zakaria95Ahmed) [reply product-images/ffa7601b-7584-47ee-9aff-9be0cee04940.png:1]: you are right i will change it

File: src/main/java/com/fruits/ecommerce/configuration/SecurityConfig/JWT_Filters/JwtAuthorizationFilter.java (hunk 1 above)
Comment (Zakaria95Ahmed) [reply src/main/java/com/fruits/ecommerce/configuration/SecurityConfig/JWT_Filters/JwtAuthorizationFilter.java:34]: you are right i will change it

File: src/main/java/com/fruits/ecommerce/configuration/SecurityConfig/SecurityCore/UserAuthorizationUtils.java (hunk 2 above)
Comment (Zakaria95Ahmed) [reply src/main/java/com/fruits/ecommerce/configuration/SecurityConfig/SecurityCore/UserAuthorizationUtils.java:8]: you are right i will change it

File: src/main/java/com/fruits/ecommerce/controller/CartController.java (hunk 4 above)
Comment (Zakaria95Ahmed) [reply src/main/java/com/fruits/ecommerce/controller/CartController.java:30]: you are right i will change it

File: src/main/java/com/fruits/ecommerce/models/dtos/CartItemDTO.java (hunk 5 above)
Comment (Zakaria95Ahmed) [reply src/main/java/com/fruits/ecommerce/models/dtos/CartItemDTO.java:11]: ok,i will change it as u want

File: src/main/java/com/fruits/ecommerce/services/Interfaces/IUserService.java (hunk 7 above)
Comment (Zakaria95Ahmed) [reply src/main/java/com/fruits/ecommerce/services/Interfaces/IUserService.java:48]: Ok , done before

File: src/main/resources/db/migration/V3__Create_products_and_product_images_tables.sql (hunk 9 above)
Comment (Zakaria95Ahmed) [reply src/main/resources/db/migration/V3__Create_products_and_product_images_tables.sql:35]: ok,i will change it

**Pull request merged**